### PR TITLE
Start reconnect transport.

### PR DESCRIPTION
**Pull request by floitsch**

### CODE REVIEW

File: src/client.toit [hunk 1]
```diff
@@ -219,13 +226,19 @@ class Client:
...
       remaining_keep_alive_us := keep_alive_.in_us - (Time.monotonic_us - last_sent_us_)
-      packet := ?
+      packet := null
       if remaining_keep_alive_us <= 0:
         packet = null
       else:
         remaining_keep_alive := Duration --us=remaining_keep_alive_us
         // Timeout returns a `null` packet.
-        packet = transport_.receive --timeout=remaining_keep_alive
+        exception := catch:
```
Comment (kasperl) [inline src/client.toit:235]: Maybe you should be using `--trace=(: should_trace_exception it)` here?

File: src/client.toit [hunk 2]
```diff
@@ -205,7 +205,14 @@ class Client:
     // is critical that the packet bits sent over the transport stream
     // aren't interleaved, so we use a mutex to serialize the sends.
     sending_.do:
-      transport_.send packet
+      exception := catch:
```
Comment (kasperl) [inline src/client.toit:208]: Maybe you should be using `--trace=(: should_trace_exception it)` here?

File: src/tcp.toit [hunk 3]
```diff
@@ -31,3 +33,84 @@ class TcpTransport implements Transport:
...
+        // TODO(florian): implement exponential back-off.
+        transport_ = new_transport_
+      finally:
+        reconnecting_mutex = null
+
+  new_connection_ -> tcp.Socket:
+    socket := interface_.tcp_connect host_ port_
+    if use_tls_:
+      socket = tls.Socket.client socket
+        --server_name=server_name_ or host_
```
Comment (kasperl) [inline src/tcp.toit:103]: 4 space indent.

File: src/tcp.toit [hunk 4]
```diff
@@ -31,3 +33,84 @@ class TcpTransport implements Transport:
...
+          close_transport_
+
+        // TODO(florian): implement exponential back-off.
+        transport_ = new_transport_
+      finally:
+        reconnecting_mutex = null
+
+  new_connection_ -> tcp.Socket:
+    socket := interface_.tcp_connect host_ port_
+    if use_tls_:
```
Comment (kasperl) [inline src/tcp.toit:101]: Maybe negate this check?

File: src/transport.toit [hunk 5]
```diff
@@ -11,11 +11,20 @@ interface Transport:
...
   Receive the next packet from the peer.
 
   Returns null if timeout was exceeded.
   */
   receive --timeout/Duration?=null -> Packet?
+
+/**
+A transport that can reconnect in case of failures.
+*/
+interface ReconnectTransport extends Transport:
```
Comment (kasperl) [inline src/transport.toit:26]: ReconnectingTransport?

File: src/tcp.toit [hunk 6]
```diff
@@ -31,3 +33,84 @@ class TcpTransport implements Transport:
...
+    transport_ = new_transport_
+
+  send packet/Packet:
+    transport_.send packet
+
+  receive --timeout/Duration?=null -> Packet?:
+    return transport_.receive --timeout=timeout
+
+  reconnect:
+    if reconnecting_mutex:
```
Comment (kasperl) [inline src/tcp.toit:85]: It feels a little bit odd to use a mutex for this. Maybe a latch would be more appropriate? I know you don't need a value to flow through.

An alternative would be something where you don't change the mutex all the time:

```
existing := transport_
reconnecting_mutex.do:
  transport := transport_
  if not (identical existing transport): return
  if transport: close_transport_
  transport_ = new_transport_
```

File: src/tcp.toit [hunk 7]
```diff
@@ -31,3 +33,84 @@ class TcpTransport implements Transport:
       with_timeout timeout:
         return Packet.deserialize reader_
     return null
+
+/**
+A transport for backing an MQTT client with TCP or TLS/TCP.
+
+Supports reconnecting to the same server.
+*/
+class ReconnectTcpTransport implements ReconnectTransport:
```
Comment (kasperl) [inline src/tcp.toit:42]: Maybe the TLS and non-TLS versions should be split into two classes?

**Review by kasperl — COMMENTED** [review]
I think this looks like the right direction!

File: src/tcp.toit (hunk 3 above)
Comment (floitsch) [reply src/tcp.toit]: not there anymore.

File: src/tcp.toit (hunk 4 above)
Comment (floitsch) [reply src/tcp.toit]: different class now.

File: src/tcp.toit (hunk 6 above)
Comment (floitsch) [reply src/tcp.toit]: I considered using a latch, but the comment states there should be at most one call to `get`.
It's probably true that there will only ever be two tasks that will try to reconnect, but I didn't want to rely on it.

Switched to what you proposed.

File: src/tcp.toit (hunk 7 above)
Comment (floitsch) [reply src/tcp.toit]: done.

File: src/client.toit (hunk 1 above)
Comment (floitsch) [reply src/client.toit]: done.

File: src/client.toit (hunk 2 above)
Comment (floitsch) [reply src/client.toit]: done.

File: src/client.toit [hunk 8]
```diff
@@ -219,13 +226,17 @@ class Client:
...
         remaining_keep_alive := Duration --us=remaining_keep_alive_us
         // Timeout returns a `null` packet.
-        packet = transport_.receive --timeout=remaining_keep_alive
+        exception := catch --trace=(: should_trace_exception_ it):
+          packet = transport_.receive --timeout=remaining_keep_alive
+        if exception:
+          if is_closed: return
+          if transport_ is ReconnectingTransport:
+            (transport_ as ReconnectingTransport).reconnect
+            continue
```
Comment (kasperl) [inline src/client.toit:239]: Here we retry multiple times. On send, we only try to reconnect once.

File: src/tcp.toit [hunk 9]
```diff
@@ -31,3 +93,54 @@ class TcpTransport implements Transport:
...
+    return interface_.tcp_connect host_ port_
+
+  close:
+    if socket_: socket_.close
+    socket_ = null
+    writer_ = null
+    reader_ = null
+
+
+class ReconnectingTlsTcpTransport_ extends ReconnectingTcpTransport_:
```
Comment (kasperl) [inline src/tcp.toit:125]: Maybe dro the Tcp part of the name? It is a local name and it doesn't add much to have it here.

**Review by kasperl — APPROVED** [review]
Looks good to me.

**Comment by floitsch** [comment]
abandoning in favor of smaller PRs.

**Pull request closed without merging**